Add state overrides

cmenapace opened the pull request:
This pull requests adds a State overwrite function to the business process.

every state that a service could have. can be translated in another service state in the business process

So you can decide for ex. to set the state of a service to warning in the business process if the state goes to critical

![1](https://user-images.githubusercontent.com/49369365/56041601-69bca300-5d39-11e9-9cad-f22babb9cb99.png)

In the Tile view you see the service with the translated state and a small circle showing the real state if differs

![2](https://user-images.githubusercontent.com/49369365/56041900-1f87f180-5d3a-11e9-9a00-d116cf2f6989.png)
![icingastate](https://user-images.githubusercontent.com/49369365/56041919-2b73b380-5d3a-11e9-8209-55d2f3d2f875.png)

In the Tree view you also see the translation state settings

![3](https://user-images.githubusercontent.com/49369365/56042034-6b3a9b00-5d3a-11e9-8815-a2d4d5b86b7d.png)

Please let me know if there are things that need to be changed.

resolves #233

nilmerg commented:
Please create an issue as well. I'd like to discuss this first. And please already include an example why such functionality may be of interest to other users.

cmenapace commented:
I opened The Issue #233

Thomas-Gelf commented:
Nice :-) Thanks for your efforts, @cmenapace, this looks already very promising. @nilmerg: I discussed this feature with @cmenapace. He has been using a variant of this, with a specific set of named state transformations. After discussing the feature with @lippserd, I suggested him to change this in a way allowing to override states in a completely custom way.

I'd love to see a short title/introduction above the "State override" section. The single states should preferably not be visible immediately. This could be an auto-submitting checkbox or dropdown, reading "Override State", defaulting "no". As soon as you switch this on, the mappings should be available, they should be removed all at once as soon as you switch this off.

@nilmerg: it would be great if you could help @cmenapace to tweak this the way you need it!

Thanks,
Thomas

cmenapace commented:
I am very happy to hear this :-). I will implement the changes asap

Thanks,
Christian

cmenapace commented:
Could this be ok

![image](https://user-images.githubusercontent.com/49369365/56281665-811ad800-610d-11e9-8504-a6cfbed7418c.png)
![image](https://user-images.githubusercontent.com/49369365/56281674-85df8c00-610d-11e9-9697-85667e6b98b2.png)

nilmerg commented on application/forms/AddNodeForm.php:
```diff
@@ -391,6 +438,34 @@ protected function enumServiceList($host)
+                }
+                $stateString .= $state. '-'. $this->getValue($stateName);
+            }
+        }
```
`==` vs. `===` applies here as well: `!==`

That's especially important if checking for `null`.

Also, please put a space before and after any operator. In this case the `.`s on line no. 462. (Though not the `.=`, that's fine)

cmenapace replied:
ok

nilmerg commented on application/forms/AddNodeForm.php:
```diff
@@ -464,8 +539,18 @@ public function onSuccess()
         switch ($this->getValue('node_type')) {
-            case 'host':
             case 'service':
+                if ($this->statesToString() != null) {
```
`!==` :wink:

cmenapace replied:
I Raplaced it with:

```suggestion
                if ($this->getValue($stateName) !== null && $this->getValue($stateName) !== '') {
```
beacase, if not set it returns an emty string, but for the OK => 0 state i dont creata an "Override State" Dropdown menu. So that returns `null`

nilmerg commented on application/forms/AddNodeForm.php:
```diff
@@ -464,8 +539,18 @@ public function onSuccess()
-            case 'host':
             case 'service':
+                if ($this->statesToString() != null) {
+                    $services = array();
```
`[]`

nilmerg commented on application/forms/AddNodeForm.php:
```diff
@@ -464,8 +539,18 @@ public function onSuccess()
+                if ($this->statesToString() != null) {
+                    $services = array();
+                    foreach ($this->getValue('children') as $service) {
+                        $services[] = $service. ':'. $this->statesToString();
```
`$service . ':' . $`..

cmenapace replied:
```suggestion
                        $services[] = $service . ':' . $this->statesToString();
```

cmenapace commented on application/forms/AddNodeForm.php:
```diff
@@ -206,7 +206,12 @@ protected function selectService()
+            $this->addServicesElement($host)
+                 ->addServiceOverrideCheckbox();
+
+            if ($this->getSentValue('service_override') == 'y') {
```
ok

cmenapace commented on application/forms/AddNodeForm.php:
```diff
@@ -236,6 +241,8 @@ protected function addServicesElement($host)
             'validators'    => [[new NoDuplicateChildrenValidator($this, $this->bp, $this->parent), true]]
         ]);
+
+        return $this;
```
ok

cmenapace commented on application/forms/AddNodeForm.php:
```diff
@@ -253,6 +260,46 @@ protected function addFileElement()
 
+    protected function addServiceOverrideCheckbox()
+    {
+        $this->addBoolean('service_override', [
```
I was not sure witch to choose at the beginning. So i looked for other Icingaweb2 modules to see what was used there and copied from them.
Personally I also prefer the checkbox ;-). I replaced it

cmenapace commented on application/forms/AddNodeForm.php:
```diff
@@ -253,6 +260,46 @@ protected function addFileElement()
+    protected function addServiceOverrideCheckbox()
+    {
+        $this->addBoolean('service_override', [
+            'label'        => 'Override Service State',
```
ok

cmenapace commented on application/forms/AddNodeForm.php:
```diff
@@ -253,6 +260,46 @@ protected function addFileElement()
+
+    protected function addServiceOverrideElements()
+    {
+        $elements = array();
```
ok

cmenapace commented on application/forms/AddNodeForm.php:
```diff
@@ -253,6 +260,46 @@ protected function addFileElement()
+            'label'        => 'Override Service State',
+            'ignore'       => true,
+            'class'        => 'autosubmit',
+            'description'   => $this->translate('Enables to override the service states'),
```
Yes. I replaced it

cmenapace commented on application/forms/AddNodeForm.php:
```diff
@@ -253,6 +260,46 @@ protected function addFileElement()
+        }
+
+        $this->addSimpleDisplayGroup($elements, 'override_group', [
+            'legend' => $this->translate('States override')
```
ok

cmenapace commented on application/forms/AddNodeForm.php:
```diff
@@ -253,6 +260,46 @@ protected function addFileElement()
+    {
+        $elements = array();
+        foreach ($this->enumServiceStateList() as $state => $stateName) {
+            if ($state == 0) {
```
```suggestion
            if ($state === 0) {
```

cmenapace commented on application/forms/AddNodeForm.php:
```diff
@@ -253,6 +260,46 @@ protected function addFileElement()
+            if ($state == 0) {
+                continue;
+            }
+            $this->addElement('select', $stateName, array(
```
$this->addElement('select', $stateName, [

cmenapace commented:
Thank you for checking the code.
I fixed the issues in the Class. I added the details in the change requests.
I have also fixed the same issues in all the other classes.

I addition I have also changed to the new array syntax here 

```
  protected function enumServiceStateList()
    {
        $serviceStateList = [
            0 => $this->translate('OK'),
            1 => $this->translate('WARNING'),
            2 => $this->translate('CRITICAL'),
            3 => $this->translate('UNKNOWN'),
            99 => $this->translate('PENDING'),
        ];
```

nilmerg commented:
Thanks for the update. Though, I won't find enough time to review this further anytime soon. I'll get back to it in a few weeks. :smiley:

sebek72 commented:
Very nice feature. I would need something similar and was wondering if this feature can be used for my case?

Two identical servers, both ok, but one is exec, one standby. The exec color should be green and standby should be shown in a different color (yellow - as ok, but standby). Exec-standby status is part of a check executed on both hosts.

cmenapace commented:
Hello, any news on that pull request?

cmenapace commented:
Hello Johannes,

I am very happy to hear this :-),

I now, I saw the big improvements and features that you guys are preparing for the new releases, especially the Icinga DB, we are really exited on that.

Back to our service overwrite

That looks great, I installed the new version in our environment and it works perfectly.
Thank you for reviewing and improving everything.
